### PR TITLE
Modify rule S6466: Extend LaYC content

### PR DESCRIPTION
**Pull request by anton-haubner-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S6466/java/rule.adoc [hunk 1]
```diff
@@ -1,56 +1,133 @@
...
+in `ArrayIndexOutOfBounds` exceptions.
+These situations can be avoided by carefully considering the range of valid
+index values, or even better, by comparing indices against the size of a
+sequence.
 
-=== Code examples
+In this first example, we create an array `arr` containing three elements.
+Since the first element of a list has index `0`, the last valid index is `2`.
+Therefore, an `ArrayIndexOutOfBoundsException` is thrown when accessing `arr` at
+index `3`.
```
Comment (Swalkyn) [inline rules/S6466/java/rule.adoc:32]: We should avoid 1st person pronouns.
```suggestion
In this first example, the array `arr` contains three elements.
The biggest valid index is `2`.
Therefore, an `ArrayIndexOutOfBoundsException` is thrown when accessing `arr` at
index `3`.
```

File: rules/S6466/java/rule.adoc [hunk 2]
```diff
@@ -1,56 +1,133 @@
...
 ----
-  void nonCompliant() {
-    int[] l = {1, 2, 3};
+  void fun() {
+    int[] arr = {1, 2, 3};
 
-    int x = l[-1]; // Noncompliant
-    int y = l[3]; // Noncompliant
+    int x = arr[3]; // Noncompliant: Valid indices are 0, 1 and 2
+    int y = arr[-1]; // Noncompliant: Negative indices are not supported in Java
```
Comment (Swalkyn) [inline rules/S6466/java/rule.adoc:42]: I would remove the comment about negative indexes, as I don't think thy are a common language feature

File: rules/S6466/java/rule.adoc [hunk 3]
```diff
@@ -1,56 +1,133 @@
...
-    int x = l[0];
-    int y = l[2];
+    int x = arr[0];
+    int y = arr[2];
   }
 ----
 
+The following example is similar, but we don't know the concrete length of the
+array `arr` so we use `array.length` instead.
+Still, accessing an array with its size as an index is not correct:
```
Comment (Swalkyn) [inline rules/S6466/java/rule.adoc:60]: I would simplify 
```suggestion
Accessing an array with its size as an index is never correct:
```

File: rules/S6466/python/rule.adoc [hunk 4]
```diff
@@ -26,11 +58,87 @@ def fun():
 
 ----
 
+The following example is similar, but we don't know the length of the list `ls`
+so it is computed using `len(...)`.
```
Comment (Swalkyn) [inline rules/S6466/python/rule.adoc:62]: See my remark for the corresponding example in Java

File: rules/S6466/python/rule.adoc [hunk 5]
```diff
@@ -26,11 +58,87 @@ def fun():
...
+    foo(i, elem)
+----
+
+==== Compliant solution
+
+[source,python,diff-id=3,diff-type=compliant]
+----
+for i, elem in enumerate(ls):
+  foo(i, elem)
+----
```
Comment (Swalkyn) [inline rules/S6466/python/rule.adoc:133]: I like how you wrote this section! I think it's concise and useful.

**Review by Swalkyn — APPROVED** [review]
Nice job! The format looks good and I've left some suggestions on the content. You can decide if you want to include them or not.

File: rules/S6466/java/rule.adoc [hunk 6]
```diff
@@ -1,56 +1,133 @@
 
-An array index out of bounds exception is a bug class that occurs in Java when a program tries to access an array element that does not exist. This bug can cause a Java program to crash or behave unexpectedly, and it can also lead to security vulnerabilities. To fix an array index out of bounds exception in Java, you should always ensure that you are accessing array elements within the bounds of the array.
+An array index out-of-bounds exception is a bug class that occurs in Java when a program tries to access an array element that does not exist.
+This bug can cause your program to crash or behave unexpectedly, and it can also lead to security vulnerabilities.
```
[the remark this answers is not in the document]
Comment (anton-haubner-sonarsource) [reply rules/S6466/java/rule.adoc:3]: You are right, its not the same as for C. The comment about security vulnerabilities was already part of the original description and I am also on the fence about it...

Although availability/reliability can be security concerns... I would say I remove the term "security vulnerabilities" here, but keep some of the comments in `Potential Impacts`.

File: rules/S6466/java/rule.adoc (hunk 2 above)
Comment (anton-haubner-sonarsource) [reply rules/S6466/java/rule.adoc:42]: They are a feature if you are coming from Python, but I see what you mean. I'll remove it, its enough to mention it in the surrounding text that negative indices are not allowed.

**Pull request merged**